### PR TITLE
Make sure circular contours don't throw a warning

### LINKED ISSUE #8529: clabel throws runtime error for circular-like contours
The following code produces a Runtime Warning thrown by `clabel`. I think it has to do with the circular-like contours. I often produce contours that look like this and keep getting this Runtime Warning, and I narrowed it down to this example. Is there anything we can fix in `clabel`? matplotlib version 2.0.0

```
import numpy as np
import matplotlib.pyplot as plt

x, y = np.meshgrid(np.linspace(-2, 2, 4), np.linspace(-2, 2, 4))
r = np.sqrt(x ** 2 + y ** 2)
plt.figure()
cs = plt.contour(x, y, r)
plt.clabel(cs)
```

### PR DESCRIPTION
**Pull request by dstansby**
Fixes #8529. This feels very hacky though - is there a better way to make division by zero return np.inf and not raise a warning?

### CODE REVIEW

**Comment by jenshnielsen** [comment]
You can numpy.errstate to change how this is handled. https://docs.scipy.org/doc/numpy/reference/generated/numpy.errstate.html

```
with np.errstate(divide='ignore', invalid='ignore'):
    b = np.arange(10)/0
```

**Comment by dstansby** [comment]
Hmm, looks like the windows build with python 2.7 and numpy 1.8 is throwing 4 warnings where there should be none on my new test.

**Comment by tacaswell** [comment]
The warnings are coming from the `ceil` in line 456 not the warnings you fixed.

I am confused why none of the other configuration raise it.

Appveyor is using `numpy                     1.12.1` from the defaults channel.

**Comment by dstansby** [comment]
I've added another error ignore statement to catch that error - it seems to be raising when `np.ceil` is passed `np.nan` as a value.

File: lib/matplotlib/contour.py [hunk 1]
```diff
@@ -354,7 +355,10 @@ def locate_label(self, linecontour, labelwidth):
         xlast = XX[:, -1].reshape(xsize, 1)
         s = (yfirst - YY) * (xlast - xfirst) - (xfirst - XX) * (ylast - yfirst)
         L = np.sqrt((xlast - xfirst) ** 2 + (ylast - yfirst) ** 2).ravel()
-        dist = np.add.reduce(([(abs(s)[i] / L[i]) for i in range(xsize)]), -1)
+        # Ignore warning that divide by zero throws, as this is a valid option
+        with np.errstate(divide='ignore', invalid='ignore'):
+            dist = np.add.reduce(([(abs(s)[i] / L[i]) for i in range(xsize)]),
```
Comment (QuLogic) [inline lib/matplotlib/contour.py]: Extra parentheses.

Is this `abs(s[:, np.newaxis]) / L[np.newaxis, :]` or `abs(s[np.newaxis, :]) / L[:, np.newaxis]`?

Comment (dstansby) [inline lib/matplotlib/contour.py]: Huh? I don't think I've done anything on this line apart from move the `-1)` on to the next line

Comment (QuLogic) [inline lib/matplotlib/contour.py]: Yes, the first suggestion is just cleanup. The second is a question for my curiosity if you happen to know the answer.

Comment (dstansby) [inline lib/matplotlib/contour.py]: I have no idea - I tried to understand what was going on here for a while but gave up...

**Pull request merged**